Issue with Type inference?

The following code snippet does not compile with Dotty but does with Scalac.
```scala
object Test {
  def second[A, B <: A](foo: A, bar: B): B = bar
  def expectString(s: String) = s

  def test = {
    second(Set.empty[String], Set.empty) map expectString
  }
}
```
It fails with the following error:
```scala
./tests/allan/Experiment.scala:7: error: type mismatch:
 found   : Any(s)
 required: String
    second(Set.empty[String], Set.empty) map expectString
                                             ^
```
It also fails to compile if `second` is defined as follow:
```scala
def second[A, B >: A](foo: A, bar: B): B = bar
```
with the following error:
```scala
./tests/allan/Experiment.scala:7: error: type mismatch:
 found   : (scala.collection.immutable.Set[String] | scala.collection.immutable.Set[Any]) @UnsafeNonvariant#A(s)
 required: String
    second(Set.empty[String], Set.empty) map expectString
                                             ^
```
Note that it does compile with `second` applied on `List` instead of `Set`:
```scala
second(List.empty[String], List.empty) map expectString
```
Not sure if it is a duplicate of #998

--- a/src/dotty/tools/dotc/typer/Typer.scala
+++ b/src/dotty/tools/dotc/typer/Typer.scala
@@ -1460,7 +1460,12 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
           }
       case wtp: MethodType if !pt.isInstanceOf[SingletonType] =>
         val arity =
-          if (defn.isFunctionType(pt)) defn.functionArity(pt)
+          if (defn.isFunctionType(pt))
+            if (!isFullyDefined(pt, ForceDegree.none) && isFullyDefined(wtp, ForceDegree.none))
+              // if method type is fully defined, but expected type is not,
+              // prioritize method parameter types as parameter types of the eta-expanded closure
+              0
+            else defn.functionArity(pt)
           else if (pt eq AnyFunctionProto) wtp.paramTypes.length
           else -1
         if (arity >= 0 && !tree.symbol.isConstructor)

--- a/tests/pos/i1036.scala
+++ b/tests/pos/i1036.scala
@@ -1,0 +1,12 @@
+object Test {
+  def second[sA, sB <: sA](foo: sA, bar: sB): sB = bar
+  def third[A, B >: A](foo: A, bar: B): B = bar
+  def expectString(s: String) = s
+
+  def test = {
+    val x = second(Set.empty[String], Set.empty)
+    x map expectString
+    second(Set.empty[String], Set.empty) map expectString
+    third(Set.empty[String], Set.empty) map expectString
+  }
+}

